### PR TITLE
Kernel/aarch64: Road to Userland [5/N] (Build Panic & printf)

### PR DESCRIPTION
**Pull request by konradekk**
This merge request is based on 4/N part by @FireFox317, which (sadly for me!) he delivered first. (Thus the nod to his pattern of naming! 🎩)

Still Iʼve got something left in my source tree to share. Some of the more general work has been also extracted to #17205.

### CODE REVIEW

File: Kernel/API/Syscall.h
Comment (nico) [inline Kernel/API/Syscall.h]: This shouldn't be in a PR titled "Kernel/aarch64". Put it in a separate PR, maybe?

File: Kernel/API/Syscall.h
Comment (nico) [inline Kernel/API/Syscall.h:366]: …doesn't this change render the previous two patches moot? Maybe squash the first 3 commits together?

File: Kernel/Arch/aarch64/RPi/MMIO.cpp
Comment (nico) [inline Kernel/Arch/aarch64/RPi/MMIO.cpp:15]: Why is this better than what was here before? The code's still device-dependent (cf 0x3f00'0000), and on the LHS the checking code is device-dependent too. That's imho clearer (?)

File: Kernel/Arch/aarch64/Registers.h
Comment (nico) [inline Kernel/Arch/aarch64/Registers.h:725]: I'm not sure I understand this change. Is making these `unsigned` what helps?

Changing the size of the int type shouldn't have a big effect on the itanium abi if there's a bitfield size, I thought. Do you have a godbolt example where this makes a difference?

File: Kernel/CMakeLists.txt
Comment (nico) [inline Kernel/CMakeLists.txt:467]: Should we rm this file now?

Comment (nico) [reply Kernel/CMakeLists.txt:467]: …oh I see, it's in the next commit. Squash it in this one.

File: Kernel/Arch/aarch64/Panic.cpp
Comment (nico) [inline Kernel/Arch/aarch64/Panic.cpp:3]: i'd squash this into the commit two prior, where the need for this file is removed.

**Comment by nico** [comment]
Overall, looks great, but I think it could make do with some squashing. And I don't like the two MainIdRegister commits.

**Comment by GMTA** [comment]
Why are three commits in this PR also in PR #17201? This PR does not seem to be based on that branch.

Either base this branch on that branch and mention that PR #17201 needs to be merged first, or keep everything in one PR.

File: Kernel/API/Syscall.h
Comment (konradekk) [reply Kernel/API/Syscall.h]: Oh! Missed that completely, thanks! Iʼve messed commits across my pull requests! This changes landed in #17205, where they belong.

File: Kernel/Arch/aarch64/RPi/MMIO.cpp
Comment (konradekk) [reply Kernel/Arch/aarch64/RPi/MMIO.cpp]: This is yet another instance of me messing up the pull requests, this belongs to #17201; Iʼm pretty convinced itʼs a decent change and Iʼll gladly explain the my reasoning there! 🙃

File: Kernel/Arch/aarch64/Registers.h
Comment (konradekk) [reply Kernel/Arch/aarch64/Registers.h]: Removed from here as it belongs to #17201 (short answer: need not to use mask when reading revision—it was definitely related to the sign, as you noted, the snippet Iʼve used did not land yet in a PR, though).

File: Kernel/CMakeLists.txt
Comment (konradekk) [reply Kernel/CMakeLists.txt:467]: Yup, done!

File: Kernel/Arch/aarch64/Panic.cpp
Comment (konradekk) [reply Kernel/Arch/aarch64/Panic.cpp:3]: Sure thing; done!

**Comment by konradekk** [comment]
Messed commits across pull requests; now everyting should be in order:
* MIDR unification got out into #17201, and
* thread-related kernel changes are in #17205.

File: Kernel/API/Syscall.h
Comment (konradekk) [reply Kernel/API/Syscall.h]: Got out to #17205 – excuse me for the commotion! 🙇🏻‍♂️

**Comment by konradekk** [comment]
> Overall, looks great, but I think it could make do with some squashing. And I don't like the two MainIdRegister commits.

Removed it from this PR – ignoring my mistake, itʼll be easier to get this PR accepted without that change! 😅

**Comment by GMTA** [comment]
> Sorry for wasting your time on that, @GMTA!

No worries, this looks better, thanks :^)

File: Kernel/Arch/aarch64/RegisterState.h
Comment (FireFox317) [inline Kernel/Arch/aarch64/RegisterState.h:22]: Again the commit message (b5f36d2c09cd31de956af73994f541f7b9691fcc) does not describe what is actually changed. In any case, the word `link` does not make any sense here.

File: Kernel/Arch/aarch64/Registers.h
Comment (FireFox317) [inline Kernel/Arch/aarch64/Registers.h:697]: Commit message again (2bc99853ea6ea32f353890fb9d36e9acef9fc4c9), and I also have a commit for this in #17258 (8c7c20f153a5b79c09bbdae868faf24f4adc3f14) where I approached it differently, as we already restore `tpidr_el0` in RegisterState. I'm not sure which approach is better.

File: Kernel/Assertions.cpp
Comment (FireFox317) [inline Kernel/Assertions.cpp:3]: Do you actually have this email address?

File: Kernel/Assertions.cpp
Comment (FireFox317) [inline Kernel/Assertions.cpp:44]: Did you actually check what happens when we hit an assertion failure now? Because I think this instruction will generate a synchronous exception, which will call `exception_common`, which will choke on this exception since it is an unexpected exception, and then call `PANIC` again, ending up in an infinite loop..

File: Kernel/Arch/aarch64/Processor.h
Comment (FireFox317) [inline Kernel/Arch/aarch64/Processor.h:168]: Again the commit message (333723e544c51ab7370341515dcc09754743de3f) should describe better what is actually changed, because this commit does more than one thing.

**Review by FireFox317 — CHANGES_REQUESTED** [review]
In general the commit messages need some love, and I'm wondering if the panic code actually works as expected.

File: Kernel/Syscalls/sigaction.cpp
Comment (konradekk) [reply Kernel/Syscalls/sigaction.cpp:85]: iʼve rebased on your changes and in fact, the messages lost their sense (you can guess that i wanted to do exactly the same you did in other pull request). will clear that up! 🙇🏻‍♂️

File: Kernel/Assertions.cpp
Comment (konradekk) [reply Kernel/Assertions.cpp:3]: nope. i used it before and nobody complained. 😬

File: Kernel/Assertions.cpp
Comment (konradekk) [reply Kernel/Assertions.cpp:44]: nope; analogous instruction (triggering an exception too) is used in x86-64 code and did go blindly here—iʼd still abstract that.

would you accept making `undefined_behaviour` a to do for Aarch64 and comment why we donʼt want to have it there yet? (`__builtin_unreachable` can have similar effect, though)

File: Kernel/Arch/aarch64/Registers.h
Comment (konradekk) [reply Kernel/Arch/aarch64/Registers.h:697]: i must say iʼve liked this central point here for all the registers… and snippets also in one place.

still iʼll follow your lead here! your call?

File: Kernel/Arch/aarch64/RegisterState.h
Comment (FireFox317) [reply Kernel/Arch/aarch64/RegisterState.h:22]: The commit message still does not describe all the changes, copying the register state in fork is one of the changes part of this commit, however the other part that implements various function in RegisterState.h and ThreadRegisters.h should be a different commit, or at least should be mentioned in the commit message.

File: Kernel/Arch/aarch64/Registers.h
Comment (FireFox317) [reply Kernel/Arch/aarch64/Registers.h:697]: >  must say iʼve liked this central point here for all the registers… and snippets also in one place.

Yes I guess this is fine too, however that means we don't need to store tpidr_el0 in the RegisterState.h, but no need to add that to this PR.

Still the commit message isn't very clear, it should at least mention the TPIDR_EL0  register.

File: Kernel/Assertions.cpp
Comment (FireFox317) [reply Kernel/Assertions.cpp:44]: > nope; analogous instruction (triggering an exception too) is used in x86-64 code and did go blindly here—iʼd still abstract that.

Yes, but the x86-64 code actually handles the exception (I think somewhere in Arch/x86_64/Interrupts.cpp). The current aarch64 does not handle the `udf 0` exception, which will result in an infinite loop of exceptions, which is why you should actually test this code.

Also, I think `Processor::undefined_behaviour()` is kinda a bad name, because this will cause totally normal (defined) behaviour to be executed (which in this case is crashing the kernel).

**Review by linusg — CHANGES_REQUESTED** [review]
As in all your other PRs: please update the commit email.

File: Kernel/Arch/DebugOutput.h
Comment (kleinesfilmroellchen) [inline Kernel/Arch/DebugOutput.h:13]: Where is the purpose in this distinction if one just calls the other in all cases? As far as I can tell, kprintf is the only user of this API anyways. You can have the declaration shared, and the definition in two different cpp files for two different architectures.

**Review by kleinesfilmroellchen — CHANGES_REQUESTED** [review]
I also noticed some of @FireFox317's requests, and I have one extra thing

File: Kernel/Arch/DebugOutput.h
Comment (konradekk) [reply Kernel/Arch/DebugOutput.h:13]: you can see the difference in usage in `kprintf.cpp` which does not care for the platform it is run on.

as for the first of your points: if i understood correctly you claim that i always call `platform_debug_output` from `debug_output`? thatʼs only true for RPi (Aarch64) branch. implementation for x86-64 differs, cf. `DebugOutput` in `x86_64`.

that is still subject for a change as one would probably like to see implementations more aligned. yet i considered any more changes here as out of scope — mostly due to the fact my pull requests already tended to take considerable amount of time (mostly due to the fact they could be pretty big). 🫣

is there something i should amend or are you satisfied with my explanations and this is good as is? 😅

Comment (kleinesfilmroellchen) [reply Kernel/Arch/DebugOutput.h:13]: I'm not fully convinced but I don't want to block the PR on something that's this trivial to change.

Comment (kleinesfilmroellchen) [reply Kernel/Arch/DebugOutput.h:13]: Hmmm, I think I understand now. Consider this comment resolved :^)

Comment (konradekk) [reply Kernel/Arch/DebugOutput.h:13]: oh! 🎩

**Comment by linusg** [comment]
Closing, feel free to reopen when you decide to change your commit email to something proper. This will not get merged otherwise.

**Comment by konradekk** [comment]
sure thing; thanks! 🙇🏻‍♂️

**Pull request closed without merging**
